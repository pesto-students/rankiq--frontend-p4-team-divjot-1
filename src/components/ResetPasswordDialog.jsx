import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useController, useForm } from 'react-hook-form';
// @mui
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import Alert from '@mui/material/Alert';
import Snackbar from '@mui/material/Snackbar';
import CircularProgress from '@mui/material/CircularProgress';
// constants and services
import { REGEX } from '../constants';
import { SERVER_BASE_API, USER_API } from '../constants/endpoints'; // TODO

// eslint-disable-next-line react/prop-types
function ResetPasswordDialog({ open, handleClose }) {
  const { t } = useTranslation();
  const [error, setError] = useState(false);
  const [success, setSuccess] = useState(false);
  const [loading, setLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  const { control, reset, getValues } = useForm({
    mode: 'onChange',
    defaultValues: { email: '' },
  });

  useEffect(() => {
    reset();
  }, [open]);

  const {
    field: { ref, ...userNameProps },
    fieldState: { invalid, error: userNameError },
  } = useController({
    name: 'email',
    control,
    rules: {
      required: t('errors.required'),
      pattern: { value: REGEX.EMAIL, message: t('errors.email') },
    },
  });

  const handleResetSubmit = () => {
    setLoading(true);
    const data = {
      email: getValues('email'),
    };
    fetch(`${SERVER_BASE_API}${USER_API.RESET_PASSWORD_LINK}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }).then((res) => {
      if (!res.ok || res.status >= 400) {
        if (res.status === 404) {
          setErrorMsg(t('resetPassword.errors.noAccountFound'));
        } else {
          setErrorMsg(t('resetPassword.errors.sendResetLink'));
        }
        setError(true);
      } else {
        setSuccess(true);
        handleClose(false);
      }
      setLoading(false);
    });
  };

  return (
    <>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>{t('resetPassword.title')}</DialogTitle>
        <DialogContent>
          <DialogContentText>{t('resetPassword.subtitle')}</DialogContentText>
          <form>
            <TextField
              data-testid="ipemail"
              {...userNameProps}
              inputRef={ref}
              error={invalid}
              label={t('login.email')}
              type="email"
              helperText={userNameError?.message}
              margin="dense"
              fullWidth
              required
              variant="standard"
            />
          </form>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>{t('generic.cancel')}</Button>
          <Button
            data-testid="btnResetPwd"
            disabled={invalid || getValues('email') === '' || loading}
            onClick={handleResetSubmit}
          >
            {loading ? <CircularProgress size={20} /> : t('generic.submit')}
          </Button>
        </DialogActions>
      </Dialog>

      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        open={error}
        autoHideDuration={6000}
        onClose={() => setError(false)}
      >
        <Alert severity="error">{errorMsg}</Alert>
      </Snackbar>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        open={success}
        autoHideDuration={6000}
        onClose={() => setSuccess(false)}
      >
        <Alert severity="success">
          {t('resetPassword.resetPasswordLinkSent')}
        </Alert>
      </Snackbar>
    </>
  );
}

export default ResetPasswordDialog;
