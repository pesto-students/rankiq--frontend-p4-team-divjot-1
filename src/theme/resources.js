import enDefault from './l10n/en.json';
import hiDefault from './l10n/hi.json';

const resources = {
  en: {
    default: enDefault,
  },
  hi: {
    default: hiDefault,
  },
};

export default resources;
