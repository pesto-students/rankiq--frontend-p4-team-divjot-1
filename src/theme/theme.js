import { createTheme } from '@mui/material/styles';

const themeOptions = {
  palette: {
    type: 'light',
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#ffffff',
    },
    text: { navFooter: '#FFFFFF', subtitle: 'rgba(0,0,0,54%)' },
    box: { main: '#FAFAFA' },
  },
};

const theme = createTheme(themeOptions);

export default theme;
