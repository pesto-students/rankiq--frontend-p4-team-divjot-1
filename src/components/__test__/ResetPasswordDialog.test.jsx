import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { ThemeProvider } from '@mui/material/styles';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ResetPasswordDialog from '../ResetPasswordDialog';
import theme from '../../theme/theme';

const mockStore = configureStore([thunk]);
const navigate = vi.fn();
vi.mock('react-router-dom', async () => {
  return {
    ...(await vi.importActual('react-router-dom')),
    useNavigate: () => navigate,
  };
});

describe('Dashboard.jsx', () => {
  let tree;
  let store;
  const storeData = {
    auth: {
      userInfo: { email: 'email', firstName: 'Test Name' },
      accessToken: 'testToken',
    },
    userHistory: {
      loading: false,
      data: [
        { rollNumber: '1234', mark: 50 },
        { rollNumber: '4567', mark: 11 },
      ],
    },
    rankData: {
      loading: false,
      error: null,
      data: {
        overall: { rank: 1, overall: 12 },
        zone: { rank: 1, overall: 12 },
        zoneCategory: { rank: 1, overall: 12 },
        zoneShift: { rank: 1, overall: 12 },
        zoneShiftCategory: { rank: 1, overall: 12 },
        rawMarks: 12.12,
        zoneSelected: 'test zone',
      },
    },
  };
  beforeEach(() => {
    store = mockStore(storeData);
    tree = render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <BrowserRouter>
            <ResetPasswordDialog open onClose={vi.fn()} />
          </BrowserRouter>
        </ThemeProvider>
      </Provider>
    );
  });

  it('should render without any errors', () => {
    const { asFragment } = tree;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should disable login button if invalid value is entered', async () => {
    const user = userEvent.setup();
    const emailField = screen.getByTestId('ipemail');
    const emailIp = emailField.getElementsByTagName('input')[0];

    await user.click(emailIp);
    await user.keyboard('test@test.com');

    expect(screen.queryByText('errors.email')).not.toBeInTheDocument();
    expect(screen.queryByTestId('btnResetPwd')).not.toBeDisabled();

    await user.click(screen.queryByTestId('btnResetPwd'));
  });
});
