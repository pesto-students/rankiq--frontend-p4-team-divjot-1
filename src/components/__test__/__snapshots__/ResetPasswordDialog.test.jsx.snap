// Vitest Snapshot v1

exports[`Dashboard.jsx > should render without any errors 1`] = `<DocumentFragment />`;
